### PR TITLE
Schema contract testing

### PR DESCRIPTION
**Pull request by kathvish**
*Description of changes:*
Added sample tests for the pattern "Event-driven architecture - schema and contract testing" 

By submitting this pull request, I confirm that you can use, modify, copy, and redistribute this contribution, under the terms of your choice.

### CODE REVIEW

**Comment by rohanmeh** [comment]
Assigning to @brnkrygs for now, feel free to re-assign

File: typescript-test-samples/schema-and-contract-testing/.prettierrc.js [hunk 1]
```diff
@@ -0,0 +1,7 @@
+module.exports = {
+    semi: true,
+    trailingComma: "all",
+    singleQuote: true,
+    printWidth: 120,
+    tabWidth: 4
```
Comment (brnkrygs) [inline typescript-test-samples/schema-and-contract-testing/.prettierrc.js:6]: `tabWidth` 4 here is causing a lot of `prettier` errors in the `*.ts` files, which are indented with 2 spaces.

Could you either:
1. Update this config to use two spaces (preferred)
2. Update code to match prettier config here

File: typescript-test-samples/schema-and-contract-testing/schemas/Json/CustomerCreated-v1.0.0.json
Comment (brnkrygs) [inline typescript-test-samples/schema-and-contract-testing/schemas/Json/CustomerCreated-v1.0.0.json:1]: Please keep directory names lowercase for consistency `json` over `Json`

File: typescript-test-samples/schema-and-contract-testing/tests/contract-testing.test.ts [hunk 2]
```diff
@@ -0,0 +1,51 @@
+// Copyright Amazon.com, Inc. or its affiliates. All Rights Reserved.
+// SPDX-License-Identifier: MIT-0
+
+import event_1_0_0 from '../events/event-1.0.0.json';
```
Comment (brnkrygs) [inline typescript-test-samples/schema-and-contract-testing/tests/contract-testing.test.ts:4]: Getting a warning from typescript here (though the tests still run)

`Cannot find module '../schemas/Json/CustomerCreated-v1.0.0.json'. Consider using '--resolveJsonModule' to import module with '.json' extension.ts(2732)`

Could you investigate?

(this is happening for the other test file as well)

File: typescript-test-samples/schema-and-contract-testing/tests/contract-testing.test.ts [hunk 3]
```diff
@@ -0,0 +1,51 @@
...
+    return false;
+
+  const addressWords = event.address.split(",");
+  if (addressWords.length != 4)
+    return false;
+
+  return true;
+}
+
+test('business logic in the producer has not changed', async () => {
```
Comment (brnkrygs) [inline typescript-test-samples/schema-and-contract-testing/tests/contract-testing.test.ts:20]: We've been using `describe` blocks as the root container for tests to add context and help with grouping. Could we do so here as well?

File: typescript-test-samples/schema-and-contract-testing/tests/schema-testing.test.ts [hunk 4]
```diff
@@ -0,0 +1,52 @@
...
+import jsonSchemaDiff from 'json-schema-diff';
+
+// The schema can be obtained from a repository or from Amazon EventBridge Schema Registry (the latter supports only OpenAPI 3.0 and Json Schema Draft 4 versions).
+// Here we're just reading schemas from a local json file for simplicity.
+import schema_v1_0_0 from '../schemas/Json/CustomerCreated-v1.0.0.json';
+import schema_v1_1_0 from '../schemas/Json/CustomerCreated-v1.1.0.json';
+import schema_v1_2_0 from '../schemas/Json/CustomerCreated-v1.2.0.json';
+import schema_v1_3_0 from '../schemas/Json/CustomerCreated-v1.3.0.json';
+
+test('adding new optional elements is backward compatible', async () => {
```
Comment (brnkrygs) [inline typescript-test-samples/schema-and-contract-testing/tests/schema-testing.test.ts:12]: Same as above... we've been using `describe` blocks as the root container for tests to add context and help with grouping. Could we do so here as well?

File: typescript-test-samples/schema-and-contract-testing/tests/contract-testing.test.ts [hunk 5]
```diff
@@ -0,0 +1,51 @@
+// Copyright Amazon.com, Inc. or its affiliates. All Rights Reserved.
+// SPDX-License-Identifier: MIT-0
+
+import event_1_0_0 from '../events/event-1.0.0.json';
+import event_1_2_0 from '../events/event-1.2.0.json';
+import event_1_3_0 from '../events/event-1.3.0.json';
+
+// Sample business logic code - we expect to have a full address in the "address" string, e.g. "2 Park St, Sydney, NSW 2000, Australia"
```
Comment (brnkrygs) [inline typescript-test-samples/schema-and-contract-testing/tests/contract-testing.test.ts:8]: At first I thought this function was part of the test logic, but your comment here helped me understand that its intended to be business logic that we're using as a part of the contract test.

I'm glad to see this approach! I agree that the contract tests should use the business logic code for this type of check.

I'm wondering if there's a way to emphasize the fact that this is business/application code... maybe moving this out to its own module under a `/src` top-level directory?

File: typescript-test-samples/schema-and-contract-testing/README.md [hunk 6]
```diff
@@ -0,0 +1,109 @@
...
+  - [System under test](#system-under-test-sut)
+  - [Limitations](#limitations)
+- [Contract testing](#contract-testing)
+  - [Goal](#goal-1)
+  - [Description](#description-1)
+  - [System under test](#system-under-test-sut-1)
+  - [Limitations](#limitations-1)
+
+
+- [Key Files in the Project](#key-files-in-the-project)
```
Comment (brnkrygs) [inline typescript-test-samples/schema-and-contract-testing/README.md:23]: Link is broken

File: typescript-test-samples/schema-and-contract-testing/README.md [hunk 7]
```diff
@@ -0,0 +1,109 @@
...
+- [Run the tests](#run-the-unit-tests)
```
Comment (brnkrygs) [inline typescript-test-samples/schema-and-contract-testing/README.md:24]: Link is broken

File: typescript-test-samples/schema-and-contract-testing/README.md [hunk 8]
```diff
@@ -0,0 +1,109 @@
...
+
+Both of these limitations can be addressed with [Contract testing](#contract-testing).
+
+## Contract Testing
+
+### System Under Test (SUT)
+
+This test is integration service agnostic as it uses sample events generated from schema published to a schema registry. The diagram below shows one way of implementing contract testing as part of producer’s CI/CD pipeline.
+
+The diagram below shows an example of running the schema validation as part of a producer’s CI/CD pipeline. A sample unit test from the "Build" step is provided in this pattern - ![Contract Unit Test](tests/contract-testing.test.ts).
```
Comment (brnkrygs) [inline typescript-test-samples/schema-and-contract-testing/README.md:70]: Link here is broken, same as previous section.

File: typescript-test-samples/schema-and-contract-testing/README.md [hunk 9]
```diff
@@ -0,0 +1,109 @@
...
+---
+
+### Key Files in this Project
+
+  - [schema-testing.test.ts](tests/schema-testing.test.ts) - unit test for schema testing
+  - [contract-testing.test.ts](tests/contract-testing.test.ts) - unit test for contract testing
+  - [schemas](schemas) - versioned schemas used for schema testing
+  - [events](events) - sample events used for contract testing
+
+## Run the Tests
```
Comment (brnkrygs) [inline typescript-test-samples/schema-and-contract-testing/README.md:102]: Thinking we include an "install dependencies" step here, what do you think?

e.g. `npm install`

**Review by brnkrygs — CHANGES_REQUESTED** [review]
Hi @kathvish,

Thanks so much for your contribution! I'm excited to get this merged in and included in our examples.

Most of the requested changes here are stylistic, so will hopefully be a quick update. Let me know if you'd like any help, and feel free to discuss any of the requested changes in comment threads.

File: typescript-test-samples/schema-and-contract-testing/.npmignore
[the remark this answers is not in the document]
Comment (kathvish) [reply typescript-test-samples/schema-and-contract-testing/.npmignore:1]: Sure, removed.

File: typescript-test-samples/schema-and-contract-testing/.prettierrc.js (hunk 1 above)
Comment (kathvish) [reply typescript-test-samples/schema-and-contract-testing/.prettierrc.js:6]: I copied this from the sibling folders ('typescript-test-sample' and 'apigw-lambda-dynamodb') and it's using 4 spaces there. In my *.ts files there's 4 spaces for a tab as well.

File: typescript-test-samples/schema-and-contract-testing/schemas/Json/CustomerCreated-v1.0.0.json
Comment (kathvish) [reply typescript-test-samples/schema-and-contract-testing/schemas/Json/CustomerCreated-v1.0.0.json:1]: done

File: typescript-test-samples/schema-and-contract-testing/tests/contract-testing.test.ts (hunk 2 above)
Comment (kathvish) [reply typescript-test-samples/schema-and-contract-testing/tests/contract-testing.test.ts:4]: fixed

File: typescript-test-samples/schema-and-contract-testing/tests/schema-testing.test.ts (hunk 4 above)
Comment (kathvish) [reply typescript-test-samples/schema-and-contract-testing/tests/schema-testing.test.ts:12]: done

File: typescript-test-samples/schema-and-contract-testing/tests/contract-testing.test.ts (hunk 3 above)
Comment (kathvish) [reply typescript-test-samples/schema-and-contract-testing/tests/contract-testing.test.ts:20]: done

File: typescript-test-samples/schema-and-contract-testing/README.md [hunk 10]
```diff
@@ -0,0 +1,109 @@
...
+
+### System Under Test (SUT)
+
+This test is integration service agnostic, as it runs against the schema published to a schema registry. Amazon EventBridge Schema Registry is one example of such a registry. It provides schemas for all events that are generated by AWS services. When you turn on schema discovery it will automatically detect and provide schemas for your custom events (OpenAPI and JSON Schemas). You can also use your own Git repository as a registry for storing schemas. In our test we will use a local “schemas” folder as a registry for simplicity. 
+
+Schema testing checks for changes that are not backward-compatible, such as removing elements, changing a field type or adding new mandatory elements. Adding optional elements is considered backward-compatible.
+
+The diagram below shows an example of running the schema validation as part of a producer’s CI/CD pipeline. A sample unit test from the "Build" step is provided in this pattern - ![Schema Diff Unit Test](tests/schema-testing.test.ts).
```
[the remark this answers is not in the document]
Comment (kathvish) [reply typescript-test-samples/schema-and-contract-testing/README.md:51]: fixed

File: typescript-test-samples/schema-and-contract-testing/README.md (hunk 6 above)
Comment (kathvish) [reply typescript-test-samples/schema-and-contract-testing/README.md:23]: fixed. I swear I tested all the links, maybe it was a last-minute change :)

File: typescript-test-samples/schema-and-contract-testing/README.md (hunk 7 above)
Comment (kathvish) [reply typescript-test-samples/schema-and-contract-testing/README.md:24]: fixed

File: typescript-test-samples/schema-and-contract-testing/README.md (hunk 8 above)
Comment (kathvish) [reply typescript-test-samples/schema-and-contract-testing/README.md:70]: fixed

File: typescript-test-samples/schema-and-contract-testing/README.md (hunk 9 above)
Comment (kathvish) [reply typescript-test-samples/schema-and-contract-testing/README.md:102]: Added

File: typescript-test-samples/schema-and-contract-testing/README.md [hunk 11]
```diff
@@ -0,0 +1,109 @@
...
+
+With a small number of producers and consumers, it might be easy to collaborate to avoid breaking changes. But as time goes on your event-driven architectures will grow and keeping on top of producer and consumer relationships can be hard. Hence it is important to automate schema validation to prevent breaking changes. 
+
+### System Under Test (SUT)
+
+This test is integration service agnostic, as it runs against the schema published to a schema registry. Amazon EventBridge Schema Registry is one example of such a registry. It provides schemas for all events that are generated by AWS services. When you turn on schema discovery it will automatically detect and provide schemas for your custom events (OpenAPI and JSON Schemas). You can also use your own Git repository as a registry for storing schemas. In our test we will use a local “schemas” folder as a registry for simplicity. 
+
+Schema testing checks for changes that are not backward-compatible, such as removing elements, changing a field type or adding new mandatory elements. Adding optional elements is considered backward-compatible.
```
[the remark this answers is not in the document]
Comment (kathvish) [reply typescript-test-samples/schema-and-contract-testing/README.md:49]: Added

File: typescript-test-samples/schema-and-contract-testing/tests/contract-testing.test.ts (hunk 5 above)
Comment (kathvish) [reply typescript-test-samples/schema-and-contract-testing/tests/contract-testing.test.ts:8]: Makes sense. Updated this.

File: typescript-test-samples/schema-and-contract-testing/.prettierrc.js (hunk 1 above)
Comment (brnkrygs) [reply typescript-test-samples/schema-and-contract-testing/.prettierrc.js]: Ok. We can't rule out a mis-configuration on my side either I suppose. Will take a look in some other editors validate what I'm getting.

File: typescript-test-samples/schema-and-contract-testing/tests/contract-testing.test.ts [hunk 12]
```diff
@@ -0,0 +1,42 @@
+// Copyright Amazon.com, Inc. or its affiliates. All Rights Reserved.
+// SPDX-License-Identifier: MIT-0
+
+import event_1_0_0 from '../events/event-1.0.0.json';
+import event_1_2_0 from '../events/event-1.2.0.json';
+import event_1_3_0 from '../events/event-1.3.0.json';
+
+import { isAddressValid } from '../src/address-validation';
+
+describe('Conract testing examples', () => {
```
Comment (brnkrygs) [inline typescript-test-samples/schema-and-contract-testing/tests/contract-testing.test.ts:10]: Spelling error for `describe` text (I think?)

File: typescript-test-samples/schema-and-contract-testing/schemas/Json/CustomerCreated-v1.0.0.json
Comment (brnkrygs) [reply typescript-test-samples/schema-and-contract-testing/schemas/Json/CustomerCreated-v1.0.0.json:1]: I'm not sure this change took - I re-pulled and the `Json` directory is still mixed-case.

Comment (kathvish) [reply typescript-test-samples/schema-and-contract-testing/schemas/Json/CustomerCreated-v1.0.0.json:1]: I thought I only needed to change the directory name, not the files inside it too. Updated now.

File: typescript-test-samples/schema-and-contract-testing/tests/contract-testing.test.ts (hunk 12 above)
Comment (kathvish) [reply typescript-test-samples/schema-and-contract-testing/tests/contract-testing.test.ts:10]: Good catch! Fixed.

File: typescript-test-samples/schema-and-contract-testing/schemas/Json/CustomerCreated-v1.0.0.json
Comment (brnkrygs) [reply typescript-test-samples/schema-and-contract-testing/schemas/Json/CustomerCreated-v1.0.0.json:1]: I just meant the directory... `git` might've needed the file change to pick up the directory change. Looks great now, thanks.

**Pull request merged**